lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1689,8 +1683,131 @@ void swift::performAbstractFuncDeclDiagnostics(TypeChecker &TC,
+    auto *Cond = FS->getCond().getPtrOrNull();
+    if (!Cond)
+      return;
+    auto callExpr = dyn_cast<CallExpr>(Cond);
+    if (!callExpr)
+      return;
+    auto dotSyntaxExpr = dyn_cast<DotSyntaxCallExpr>(callExpr->getFn());
+    if (!dotSyntaxExpr)
+      return;
+    auto binaryExpr = dyn_cast<BinaryExpr>(dotSyntaxExpr->getBase());
+    if (!binaryExpr)
+      return;
+    auto binaryFuncExpr = dyn_cast<DeclRefExpr>(binaryExpr->getFn());
+    if (!binaryFuncExpr)
+      return;
+    auto comparisonOpName = binaryFuncExpr->getDecl()->getNameStr();
+    if (comparisonOpName != "!=" && comparisonOpName != "<")
+      return;
+    auto args = binaryExpr->getArg()->getElements();
+    auto loadExpr = dyn_cast<LoadExpr>(args[0]);
+    if (!loadExpr)
+      return;
+    auto declRefExpr = dyn_cast<DeclRefExpr>(loadExpr->getSubExpr());
+    if (!declRefExpr)
+      return;
+    if (declRefExpr->getDecl() != loopVar)
+      return;
+    endValue = args[1];
+    
+    // Verify that the increment is a simple +1 to the loop variable.
+    auto *Increment = FS->getIncrement().getPtrOrNull();
+    if (!Increment)
+      return;
+    ApplyExpr *unaryExpr = dyn_cast<PrefixUnaryExpr>(Increment);
+    if (!unaryExpr)
+      unaryExpr = dyn_cast<PostfixUnaryExpr>(Increment);
+    if (!unaryExpr)
+      return;
+    auto inoutExpr = dyn_cast<InOutExpr>(unaryExpr->getArg());
+    if (!inoutExpr)
+      return;
+    auto incrementDeclRefExpr = dyn_cast<DeclRefExpr>(inoutExpr->getSubExpr());
+    if (!incrementDeclRefExpr)
+      return;
+    auto unaryFuncExpr = dyn_cast<DeclRefExpr>(unaryExpr->getFn());
+    if (!unaryFuncExpr)
+      return;
+    if (unaryFuncExpr->getDecl()->getNameStr() != "++")
+      return;
+    strideByOne = declRefExpr->getDecl() == loopVar;
+    
+    if (!loopVar || !startValue || !endValue || !strideByOne)
+        return;
+    
+    // Verify that the loop variable is invariant inside the body.
+    VarDeclUsageChecker checker(TC, loopVar);
+    checker.suppressDiagnostics();
+    FS->getBody()->walk(checker);
+    
+    if (checker.isVarDeclEverWritten(loopVar)) {
+      diagnostic.flush();
+      TC.diagnose(FS->getStartLoc(), diag::cant_fix_c_style_for_stmt);
+    } else {
+      // There is no saved SourceLoc for the equals in a pattern initializer, unfortunately.
+      int loopVarLen = loopVar->getName().getLength();
+      SourceLoc startFind = loopVar->getStartLoc().getAdvancedLoc(loopVarLen);
+      auto str = CharSourceRange(TC.Context.SourceMgr, startFind, startValue->getStartLoc()).str();
+      auto index = str.rfind("=");
+      if (index == StringRef::npos)
+        return; // how did we find an initializer for the startValue but no equals?
+      if (index > 0 && str[index-1] == ' ')
+        index--; // eat one space before the equals, if there is one, because our replace starts with space.
+      SourceLoc equalsLoc = startFind.getAdvancedLoc(index);
```

Comment:
If the location of the "=" isn't available in the AST, please add it to the AST so we can query it directly: we'd rather not have to scan through the source code to find locations that we need after-the-fact.